nix/open-webui.nix

```diff
@@ -0,0 +1,106 @@
+# Based on: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/open-webui.nix
+{ pkgs, lib, name, config, ... }:
+let
+  inherit (lib) types;
+in
+{
+  options = {
+    enable = lib.mkEnableOption "Open-WebUI server";
+    package = lib.mkPackageOption pkgs "open-webui" { };
+
+    dataDir = lib.mkOption {
+      type = types.str;
+      default = "./data/${name}";
+      description = "The Open-WebUI data directory";
+    };
+
+    host = lib.mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      example = "0.0.0.0";
+      description = ''
+        The host address which the Open-WebUI server HTTP interface listens to.
+      '';
+    };
+
+    port = lib.mkOption {
+      type = types.port;
+      default = 1111;
+      example = 11111;
+      description = ''
+        Which port the Open-WebUI server listens to.
+      '';
+    };
+
+    environment = lib.mkOption {
+      type = types.attrsOf types.str;
+      default = {
+        SCARF_NO_ANALYTICS = "True";
+        DO_NOT_TRACK = "True";
+        ANONYMIZED_TELEMETRY = "False";
+      };
+      example = ''
+        {
+          OLLAMA_API_BASE_URL = "http://127.0.0.1:11434";
+          # Disable authentication
+          WEBUI_AUTH = "False";
+        }
+      '';
+      description = "Extra environment variables for Open-WebUI";
+    };
+
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default = {
+        processes = {
+          "${name}" =
+            let
+              setupStateDirs = lib.concatMapStrings
+                (stateDir:
+                  ''
+                    if [ ! -d "''$${stateDir}" ]; then
+                      mkdir -p "''$${stateDir}"
+                    fi
+
+                    ${stateDir}=$(readlink -f "''$${stateDir}")
+
+                    export ${stateDir}
+                  '') [ "DATA_DIR" "STATIC_DIR" "HF_HOME" "SENTENCE_TRANSFORMERS_HOME" ];
+            in
+
+            {
+              environment = {
+                DATA_DIR = config.dataDir;
+                STATIC_DIR = config.dataDir;
+                HF_HOME = config.dataDir;
+                SENTENCE_TRANSFORMERS_HOME = config.dataDir;
+              } // config.environment;
+
+              command = pkgs.writeShellApplication {
+                name = "open-webui-wrapper";
+                runtimeInputs = [ pkgs.open-webui ];
+                text = ''
+                  ${setupStateDirs}
+
+                  open-webui serve --host ${config.host} --port ${builtins.toString config.port}
```

Comment:
```suggestion
                  ${lib.getExe config.package} serve --host ${config.host} --port ${builtins.toString config.port}
```

Reply:
In `writeShellApplication` we pass the runtime dependencies using `runtimeInputs` and that’s how we can expect the executable to be present in the environment. Do you see any problems in doing so?

Reply:
If by any chance, there's another `open-webui` in the path, this might be a problem indeed. Usually it's safer to do as advised.

Reply:
> If by any chance, there's another `open-webui` in the path, this might be a problem indeed. 

Is it a problem in practice? Doesn't `writeShellApplication` prepend the runtimeInputs to `$PATH` anyway?

> This pattern is also used in `nixpkgs`.

Could you provide an example?

If we are using `lib.getExe`, we might as well remove `runtimeInputs` entirely.

Reply:
I don't know by heart, ...but... just a simple search in `nixos/modules/services`:

![image](https://github.com/juspay/services-flake/assets/252042/e7eb42d2-4088-41ff-9d64-b868a3ade375)

Is it enough ? :)

Reply:
But none of them use `writeShellApplication`? 

I can see why in systemd's `ExecStart` we have to use `lib.getExe`, since there's no `runtimeInputs` that manipulates `$PATH` before running the script/command anyway.

Reply:
That's true indeed, I overlooked that detail. In that case, I let you decide what to do here since my argument isn't totally true.

Reply:
I personally prefer `runtimeInputs` and not using `lib.getExe` in `writeShellApplication.

If we are however using `lib.getExe`, I wouldn't use `runtimeInputs`. 

The latter is probably safer, since it is possible to accidentally use the wrong package in `runtimeInputs` with `text` using an executable that is already in user's path.

Reply:
As you prefer...

Reply:
Going with the safer option then.